File: authentication-and-authorization/access-control/role_based_access_control.py
# Placeholder content for role_based_access_control.py
class Role:
    def __init__(self, name):
        self.name = name
        self.permissions = set()

    def add_permission(self, permission):
        self.permissions.add(permission)

    def has_permission(self, permission):
        return permission in self.permissions


class User:
    def __init__(self, username):
        self.username = username
        self.roles = []

    def add_role(self, role):
        self.roles.append(role)

    def has_permission(self, permission):
        return any(role.has_permission(permission) for role in self.roles)


# Example usage
def main():
    # Define roles
    admin_role = Role('admin')
    admin_role.add_permission('create_user')
    admin_role.add_permission('delete_user')

    editor_role = Role('editor')
    editor_role.add_permission('create_post')
    editor_role.add_permission('edit_post')

    # Define users
    admin_user = User('admin_user')
    admin_user.add_role(admin_role)

    editor_user = User('editor_user')
    editor_user.add_role(editor_role)

    # Check permissions
    print(f"Admin has permission to create user: {admin_user.has_permission('create_user')}")
    print(f"Editor has permission to edit post: {editor_user.has_permission('edit_post')}")
    print(f"Editor has permission to delete user: {editor_user.has_permission('delete_user')}")

if __name__ == "__main__":
    main()

File: authentication-and-authorization/access-control/role_based_access_control.c
#include <stdio.h>
#include <string.h>

#define MAX_PERMISSIONS 5
#define MAX_ROLES 2
#define MAX_USERS 2

typedef struct {
    char name[20];
    char permissions[MAX_PERMISSIONS][20];
    int permCount;
} Role;

typedef struct {
    char username[20];
    Role roles[MAX_ROLES];
    int roleCount;
} User;

void addPermission(Role *role, const char *permission) {
    if (role->permCount < MAX_PERMISSIONS) {
        strcpy(role->permissions[role->permCount++], permission);
    }
}

void addRole(User *user, Role role) {
    if (user->roleCount < MAX_ROLES) {
        user->roles[user->roleCount++] = role;
    }
}

int hasPermission(User user, const char *permission) {
    for (int i = 0; i < user.roleCount; i++) {
        for (int j = 0; j < user.roles[i].permCount; j++) {
            if (strcmp(user.roles[i].permissions[j], permission) == 0) {
                return 1;
            }
        }
    }
    return 0;
}

int main() {
    Role adminRole = {.permCount = 0};
    strcpy(adminRole.name, "admin");

    addPermission(&adminRole, "create_user");
    addPermission(&adminRole, "delete_user");

    User adminUser = {.roleCount = 0};
    strcpy(adminUser.username, "admin");

    addRole(&adminUser, adminRole);

    printf("Does admin have permission to 'create_user'? %s\n", hasPermission(adminUser, "create_user") ? "Yes" : "No");
    printf("Does admin have permission to 'edit_post'? %s\n", hasPermission(adminUser, "edit_post") ? "Yes" : "No");

    return 0;
}
